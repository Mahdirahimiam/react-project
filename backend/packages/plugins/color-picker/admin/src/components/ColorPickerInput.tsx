import * as React from 'react';

import {
  BaseButton,
  Box,
  Field,
  FieldError,
  FieldHint,
  FieldInput,
  FieldLabel,
  Flex,
  FocusTrap,
  Popover,
  Typography,
} from '@strapi/design-system';
import { CarretDown } from '@strapi/icons';
import { HexColorPicker } from 'react-colorful';
import { useIntl, MessageDescriptor } from 'react-intl';
import styled from 'styled-components';

import { useComposedRefs } from '../hooks/useComposeRefs';
import { getTrad } from '../utils/getTrad';

const ColorPreview = styled.div`
  border-radius: 50%;
  width: 20px;
  height: 20px;
  margin-right: 10px;
  background-color: ${(props) => props.color};
  border: 1px solid rgba(0, 0, 0, 0.1);
`;

const ColorPicker = styled(HexColorPicker)`
  && {
    width: 100%;
    aspect-ratio: 1.5;
  }

  .react-colorful__pointer {
    width: ${({ theme }) => theme.spaces[3]};
    height: ${({ theme }) => theme.spaces[3]};
  }

  .react-colorful__saturation {
    border-radius: ${({ theme }) => theme.spaces[1]};
    border-bottom: none;
  }

  .react-colorful__hue {
    border-radius: 10px;
    height: ${({ theme }) => theme.spaces[3]};
    margin-top: ${({ theme }) => theme.spaces[2]};
  }
`;

const ColorPickerToggle = styled(BaseButton)`
  display: flex;
  justify-content: space-between;
  align-items: center;

  svg {
    width: ${({ theme }) => theme.spaces[2]};
    height: ${({ theme }) => theme.spaces[2]};
  }

  svg > path {
    fill: ${({ theme }) => theme.colors.neutral500};
    justify-self: flex-end;
  }
`;

const ColorPickerPopover = styled(Popover)`
  padding: ${({ theme }) => theme.spaces[2]};
  min-height: 270px;
`;

/**
 * TODO: A lot of these props should extend `FieldProps`
 */
interface ColorPickerInputProps {
  intlLabel: MessageDescriptor;
  /**
   * TODO: this should be extended from `FieldInputProps['onChange']
   * but that conflicts with it's secondary usage in `HexColorPicker`
   */
  onChange: (event: { target: { name: string; value: string; type: string } }) => void;
  attribute: { type: string; [key: string]: unknown };
  name: string;
  description?: MessageDescriptor;
  disabled?: boolean;
  error?: string;
  labelAction?: React.ReactNode;
  required?: boolean;
  value?: string;
}

export const ColorPickerInput = React.forwardRef<HTMLButtonElement, ColorPickerInputProps>(
  (
    {
      attribute,
      description,
      disabled = false,
      error,
      intlLabel,
      labelAction,
      name,
      onChange,
      required = false,
      value = '',
    },
    forwardedRef
  ) => {
    const [showColorPicker, setShowColorPicker] = React.useState(false);
    const colorPickerButtonRef = React.useRef<HTMLButtonElement>(null!);
    const { formatMessage } = useIntl();
    const color = value || '#000000';

    const handleBlur: React.FocusEventHandler<HTMLDivElement> = (e) => {
      e.preventDefault();

      if (!e.currentTarget.contains(e.relatedTarget)) {
        setShowColorPicker(false);
      }
    };

    const composedRefs = useComposedRefs(forwardedRef, colorPickerButtonRef);

    return (
      <Field
        name={name}
        id={name}
        // GenericInput calls formatMessage and returns a string for the error
        error={error}
        hint={description && formatMessage(description)}
        required={required}
      >
        <Flex direction="column" alignItems="stretch" gap={1}>
          <FieldLabel action={labelAction}>{formatMessage(intlLabel)}</FieldLabel>
          <ColorPickerToggle
            ref={composedRefs}
            aria-label={formatMessage({
              id: getTrad('color-picker.toggle.aria-label'),
              defaultMessage: 'Color picker toggle',
            })}
            aria-controls="color-picker-value"
            aria-haspopup="dialog"
            aria-expanded={showColorPicker}
            aria-disabled={disabled}
            disabled={disabled}
            onClick={() => setShowColorPicker(!showColorPicker)}
          >
            <Flex>
              <ColorPreview color={color} />
              <Typography
                style={{ textTransform: 'uppercase' }}
                textColor={value ? undefined : 'neutral600'}
                variant="omega"
              >
                {color}
              </Typography>
            </Flex>
            <CarretDown aria-hidden />
          </ColorPickerToggle>
          {showColorPicker && (
            <ColorPickerPopover
              onBlur={handleBlur}
              role="dialog"
              source={colorPickerButtonRef}
              spacing={4}
            >
              <FocusTrap onEscape={() => setShowColorPicker(false)}>
                <ColorPicker
                  color={color}
                  onChange={(hexValue) =>
                    onChange({ target: { name, value: hexValue, type: attribute.type } })
                  }
                />
                <Flex paddingTop={3} paddingLeft={4} justifyContent="flex-end">
                  <Box paddingRight={2}>
                    <Typography variant="omega" as="label" textColor="neutral600">
                      {formatMessage({
                        id: getTrad('color-picker.input.format'),
                        defaultMessage: 'HEX',
                      })}
                    </Typography>
                  </Box>
                  <FieldInput
                    id="color-picker-value"
                    aria-label={formatMessage({
                      id: getTrad('color-picker.input.aria-label'),
                      defaultMessage: 'Color picker input',
                    })}
                    style={{ textTransform: 'uppercase' }}
                    value={value}
                    placeholder="#000000"
                    onChange={onChange}
                  />
                </Flex>
              </FocusTrap>
            </ColorPickerPopover>
          )}
          <FieldHint />
          <FieldError />
        </Flex>
      </Field>
    );
  }
);
