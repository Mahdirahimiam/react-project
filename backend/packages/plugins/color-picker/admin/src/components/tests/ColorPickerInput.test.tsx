import { DesignSystemProvider } from '@strapi/design-system';
import { render as renderRTL } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { IntlProvider } from 'react-intl';

import { ColorPickerInput } from '../ColorPickerInput';

const render = () => ({
  ...renderRTL(
    <ColorPickerInput
      name="color"
      value=""
      onChange={jest.fn()}
      attribute={{
        customField: 'plugin::color-picker.color',
        pluginOptions: { i18n: { localized: true } },
        type: 'string',
      }}
      intlLabel={{ id: 'color-picker', defaultMessage: 'color-picker' }}
    />,
    {
      wrapper: ({ children }) => {
        const locale = 'en';
        return (
          <IntlProvider locale={locale} messages={{}} textComponent="span">
            <DesignSystemProvider locale={locale}>{children}</DesignSystemProvider>
          </IntlProvider>
        );
      },
    }
  ),
  user: userEvent.setup(),
});

describe('<ColorPickerInput />', () => {
  it('renders and matches the snapshot', () => {
    const { container } = render();

    expect(container).toMatchSnapshot();
  });

  it('toggles the popover', async () => {
    const { user, getByRole } = render();
    await user.click(getByRole('button', { name: 'Color picker toggle' }));

    expect(getByRole('dialog')).toBeVisible();
    expect(getByRole('slider', { name: 'Color' })).toBeVisible();
    expect(getByRole('slider', { name: 'Hue' })).toBeVisible();
    expect(getByRole('textbox', { name: 'Color picker input' })).toBeVisible();
  });
});
