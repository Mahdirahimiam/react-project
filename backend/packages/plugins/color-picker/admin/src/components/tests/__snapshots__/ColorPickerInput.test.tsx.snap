// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPickerInput /> renders and matches the snapshot 1`] = `
.c9 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 {
  font-size: 0.75rem;
  line-height: 1.33;
  font-weight: 600;
  color: #32324d;
}

.c8 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #666687;
}

.c3 {
  background: #ffffff;
  padding: 8px;
  border-radius: 4px;
  border-color: #dcdce4;
  border: 1px solid #dcdce4;
  cursor: pointer;
}

.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 4px;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  position: relative;
  outline: none;
}

.c5 > svg {
  height: 12px;
  width: 12px;
}

.c5 > svg > g,
.c5 > svg path {
  fill: #ffffff;
}

.c5[aria-disabled='true'] {
  pointer-events: none;
}

.c5:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c5:focus-visible {
  outline: none;
}

.c5:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  border-radius: 50%;
  width: 20px;
  height: 20px;
  margin-right: 10px;
  background-color: #000000;
  border: 1px solid rgba(0,0,0,0.1);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 svg {
  width: 8px;
  height: 8px;
}

.c6 svg > path {
  fill: #8e8ea9;
  justify-self: flex-end;
}

<div>
  <div
    class=""
  >
    <div
      class="c0"
    >
      <label
        class="c1 c2"
        for="color"
      >
        color-picker
      </label>
      <button
        aria-controls="color-picker-value"
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="dialog"
        aria-label="Color picker toggle"
        class="c3 c4 c5 c6"
        type="button"
      >
        <div
          class="c4"
        >
          <div
            class="c7"
            color="#000000"
          />
          <span
            class="c8"
            style="text-transform: uppercase;"
          >
            #000000
          </span>
        </div>
        <svg
          aria-hidden="true"
          fill="none"
          height="1rem"
          viewBox="0 0 14 8"
          width="1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M14 .889a.86.86 0 0 1-.26.625L7.615 7.736A.834.834 0 0 1 7 8a.834.834 0 0 1-.615-.264L.26 1.514A.861.861 0 0 1 0 .889c0-.24.087-.45.26-.625A.834.834 0 0 1 .875 0h12.25c.237 0 .442.088.615.264a.86.86 0 0 1 .26.625Z"
            fill="#32324D"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="c9"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
